src/audio/SDL_mixer.c

```diff
@@ -141,24 +140,22 @@ SDL_MixAudioFormat(Uint8 * dst, const Uint8 * src, SDL_AudioFormat format,
         {
             Sint16 src1, src2;
             int dst_sample;
-            const int max_audioval = ((1 << (16 - 1)) - 1);
-            const int min_audioval = -(1 << (16 - 1));
+            const int max_audioval = SDL_MAX_SINT16;
+            const int min_audioval = SDL_MIN_SINT16;
 
             len /= 2;
             while (len--) {
-                src1 = ((src[1]) << 8 | src[0]);
+                src1 = SDL_SwapLE16(*(Sint16 *)src);
                 ADJUST_VOLUME(src1, volume);
-                src2 = ((dst[1]) << 8 | dst[0]);
+                src2 = SDL_SwapLE16(*(Sint16 *)dst);
```

Comment:
I mentioned this in the pull request [https://github.com/libsdl-org/SDL/pull/5130](https://github.com/libsdl-org/SDL/pull/5130), but  I would like to bring attention to it here, this is dangerous and will most likely invoke undefined behavior if the circumstances is right. 

**dst** is a Uint8 pointer, if you cast it to a Uint16 pointer and then dereference it you will get alignment issues and invoke undefined behavior.

For this cast to be safe, the caller must have cast a Uint16 buffer to a Uint8 pointer, if you can't guarantee that then the cast i UB.

There is more problems, SDL_SwapLE16 and SDL_SwapBE16 will not swap anything if the native order is the same, which actually changes what the code does.

Looking at the code, the data in **dst** should always be placed in little endian format, this change will break that.

Reply:
@FredrikRosenqvist : Nope. You are wrong. Check case AUDIO_S32LSB.

Reply:
@pionere The alignment issue still exists though.

Reply:
And in the cases where current code casts the Uint8 pointer to Uint32, I would actually consider that a bug.
Looking at it a little more, SDL_MixAudioFormat is a user function and no restriction is set on the **dst** parameter. Casting this pointer to anything other than Uint8 and dereferencing is an obvious a bug.

Reply:
@FredrikRosenqvist : 
1. what alignment issue is there which does not exist in case of AUDIO_S32LSB?
2. Nope, it is not a bug. Are you a programmer?

Reply:
> @FredrikRosenqvist :
> 
> 1. what alignment issue is there which does not exist in case of AUDIO_S32LSB?
> 2. Nope, it is not a bug. Are you a programmer?

`Uint8 *stream = malloc(1024 * sizeof *stream);`
Here **stream[0]** will be aligned for any type, as the memory returned by malloc guarantees that.
And **stream[1]** however is not aligned to any type.

So it all depends on how the user passes the buffers to the function.

1. Looks like a bug also. (Anyone who uses this function probably allocates the memory and passes it directly, thus no problem.)